{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MaveDB Mapping Analysis\n",
    "\n",
    "This notebook demonstrates how data from score sets in [MaveDB](https://mavedb.org/) can be mapped to human reference sequences and represented using the [Variation Representation Specification (VRS)](https://vrs.ga4gh.org/) of the [Global Alliance for Genomics and Health (GA4GH)](https://www.ga4gh.org/), as described in \"Mapping MAVE data for use in human genomics applications\" (Arbesfeld et al). After each step of the mapping pipeline, data is saved to a local [pickle](https://docs.python.org/3/library/pickle.html) checkpoint file for easy use in later steps."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "First, initialize environment parameters to enable access to required resources:\n",
    "\n",
    "* Universal Transcript Archive (UTA): see [README](https://github.com/biocommons/uta?tab=readme-ov-file#installing-uta-locally) for setup instructions. Users with access to Docker on their local devices can use the available Docker image; otherwise, start a relatively recent (version 14+) PostgreSQL instance and add data from the available database dump.\n",
    "* SeqRepo: see [README](https://github.com/biocommons/biocommons.seqrepo?tab=readme-ov-file#requirements) for setup instructions.\n",
    "* Gene Normalizer: see [documentation](https://gene-normalizer.readthedocs.io/0.3.0-dev1/install.html) for installation instructions\n",
    "* blat: Must be available on the local PATH and executable by the user. Otherwise, its location can be set manually with the BLAT_BIN_PATH env var. See the [UCSC Genome Browser FAQ](https://genome.ucsc.edu/FAQ/FAQblat.html#blat3) for download instructions. For our experiments, we placed the binary in the same directory as these notebooks.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from os import environ\n",
    "from pathlib import Path\n",
    "\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# set external resources\n",
    "environ[\"GENE_NORM_DB_URL\"] = \"postgresql://postgres@localhost:5432/gene_normalizer\"\n",
    "environ[\"UTA_DB_URL\"] = \"postgresql://uta_admin:uta@localhost:5432/uta/uta_20210129b\"\n",
    "environ[\"BLAT_BIN_PATH\"] = str(Path(\"blat\").absolute())\n",
    "environ[\"SEQREPO_ROOT_DIR\"] = \"/usr/local/share/seqrepo/2024-02-20\"  # TODO or latest/\n",
    "environ[\"DCD_MAPPING_RESOURCES_DIR\"] = str(Path(\"./mavedb_files\").absolute())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Score Set Metadata\n",
    "\n",
    "Run the cell below to access a dataframe containing metadata for the examined MaveDB score sets. The dataframe has 209 rows, corresponding to the number of \"human\" score sets, and 7 columns:\n",
    "\n",
    "* urn: The score set identifier\n",
    "* target_sequence: The target sequence for the score set\n",
    "* target_sequence type: Is the target sequence a DNA or protein sequence\n",
    "* target: The listed target for the score set (e.g. Src catalytic domain, CXCR4)\n",
    "* assembly_id: The genome build associated with the score set\n",
    "* uniprot_id: The Uniprot ID associated with the score set, if available\n",
    "* target_type: The target type associated with the score set (e.g. Regulatory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mave_metadata_dat = pd.read_csv(\"analysis_files/mave_dat.csv\", index_col=[0])\n",
    "score_sets = mave_metadata_dat[\"urn\"].to_list()[:-5]  # TODO don't need this slice when using better mave.dat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Output Directory\n",
    "\n",
    "Output from this notebook will be stored in a directory named `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = Path(\"analyis_files\")\n",
    "path.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: MaveDB Metadata to BLAT Alignment Data\n",
    "In this section of the notebook, the target sequence for each score set is run through BLAT, allowing for genomic coordinates to be linked with the target sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate BLAT Output for each Score Set\n",
    "\n",
    "Run the cell below to generate BLAT alignment output for each examined score set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                            | 0/209 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading hg38.2bit to /Users/jss009/.local/share/dcd-mapping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                            | 0/209 [00:04<?, ?it/s]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 9\u001b[0m\n\u001b[1;32m      7\u001b[0m metadata \u001b[38;5;241m=\u001b[39m get_scoreset_metadata(ss)\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m----> 9\u001b[0m     align_results[ss] \u001b[38;5;241m=\u001b[39m \u001b[43malign\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmetadata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msilent\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43muse_cached\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m AlignmentError:\n\u001b[1;32m     11\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n",
      "File \u001b[0;32m~/code/mavemap/src/dcd_mapping/align.py:319\u001b[0m, in \u001b[0;36malign\u001b[0;34m(scoreset_metadata, silent, use_cached)\u001b[0m\n\u001b[1;32m    317\u001b[0m     blat_output \u001b[38;5;241m=\u001b[39m _get_cached_blat_output(scoreset_metadata, silent)\n\u001b[1;32m    318\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 319\u001b[0m     blat_output \u001b[38;5;241m=\u001b[39m \u001b[43m_get_blat_output\u001b[49m\u001b[43m(\u001b[49m\u001b[43mscoreset_metadata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msilent\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    320\u001b[0m match \u001b[38;5;241m=\u001b[39m _get_best_match(blat_output, scoreset_metadata)\n\u001b[1;32m    322\u001b[0m msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAlignment complete.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[0;32m~/code/mavemap/src/dcd_mapping/align.py:165\u001b[0m, in \u001b[0;36m_get_blat_output\u001b[0;34m(metadata, silent)\u001b[0m\n\u001b[1;32m    163\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    164\u001b[0m     target_args \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m--> 165\u001b[0m process_result \u001b[38;5;241m=\u001b[39m \u001b[43m_run_blat\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtarget_args\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mquery_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/dev/stdout\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msilent\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    166\u001b[0m out_file \u001b[38;5;241m=\u001b[39m _write_blat_output_tempfile(process_result)\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/code/mavemap/src/dcd_mapping/align.py:85\u001b[0m, in \u001b[0;36m_run_blat\u001b[0;34m(target_args, query_file, out_file, silent)\u001b[0m\n\u001b[1;32m     67\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run_blat\u001b[39m(\n\u001b[1;32m     68\u001b[0m     target_args: \u001b[38;5;28mstr\u001b[39m, query_file: Path, out_file: \u001b[38;5;28mstr\u001b[39m, silent: \u001b[38;5;28mbool\u001b[39m\n\u001b[1;32m     69\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m subprocess\u001b[38;5;241m.\u001b[39mCompletedProcess:\n\u001b[1;32m     70\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Execute BLAT binary with relevant params.\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \n\u001b[1;32m     72\u001b[0m \u001b[38;5;124;03m    Currently, we rely on a system-installed BLAT binary accessible in the containing\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[38;5;124;03m    :return: process result\u001b[39;00m\n\u001b[1;32m     84\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 85\u001b[0m     reference_genome_file \u001b[38;5;241m=\u001b[39m \u001b[43mget_ref_genome_file\u001b[49m\u001b[43m(\u001b[49m\u001b[43msilent\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msilent\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     86\u001b[0m     bin_name \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39menviron[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBLAT_BIN_PATH\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBLAT_BIN_PATH\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m os\u001b[38;5;241m.\u001b[39menviron \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mblat\u001b[39m\u001b[38;5;124m\"\u001b[39m  \u001b[38;5;66;03m# noqa: SIM401\u001b[39;00m\n\u001b[1;32m     87\u001b[0m     command \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mbin_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mreference_genome_file\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtarget_args\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m -minScore=20 \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mquery_file\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mout_file\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[0;32m~/code/mavemap/src/dcd_mapping/resources.py:299\u001b[0m, in \u001b[0;36mget_ref_genome_file\u001b[0;34m(silent, dcd_mapping_dir)\u001b[0m\n\u001b[1;32m    297\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m genome_file\u001b[38;5;241m.\u001b[39mexists():\n\u001b[1;32m    298\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 299\u001b[0m         \u001b[43m_http_download\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgenome_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msilent\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    300\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m requests\u001b[38;5;241m.\u001b[39mHTTPError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    301\u001b[0m         msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mHTTPError when fetching reference genome file from \u001b[39m\u001b[38;5;132;01m{\u001b[39;00murl\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[0;32m~/code/mavemap/src/dcd_mapping/resources.py:80\u001b[0m, in \u001b[0;36m_http_download\u001b[0;34m(url, out_path, silent)\u001b[0m\n\u001b[1;32m     78\u001b[0m                 progress_bar\u001b[38;5;241m.\u001b[39mupdate(\u001b[38;5;28mlen\u001b[39m(chunk))\n\u001b[1;32m     79\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 80\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mchunk\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43miter_content\u001b[49m\u001b[43m(\u001b[49m\u001b[43mchunk_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m8192\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m     81\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mchunk\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m     82\u001b[0m \u001b[43m            \u001b[49m\u001b[43mh\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite\u001b[49m\u001b[43m(\u001b[49m\u001b[43mchunk\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/mavemap/.venv/lib/python3.12/site-packages/requests/models.py:816\u001b[0m, in \u001b[0;36mResponse.iter_content.<locals>.generate\u001b[0;34m()\u001b[0m\n\u001b[1;32m    814\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    815\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 816\u001b[0m         \u001b[38;5;28;01myield from\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mstream(chunk_size, decode_content\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    817\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m ProtocolError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    818\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m ChunkedEncodingError(e)\n",
      "File \u001b[0;32m~/code/mavemap/.venv/lib/python3.12/site-packages/urllib3/response.py:628\u001b[0m, in \u001b[0;36mHTTPResponse.stream\u001b[0;34m(self, amt, decode_content)\u001b[0m\n\u001b[1;32m    626\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    627\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_fp_closed(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp):\n\u001b[0;32m--> 628\u001b[0m         data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mamt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    630\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m data:\n\u001b[1;32m    631\u001b[0m             \u001b[38;5;28;01myield\u001b[39;00m data\n",
      "File \u001b[0;32m~/code/mavemap/.venv/lib/python3.12/site-packages/urllib3/response.py:567\u001b[0m, in \u001b[0;36mHTTPResponse.read\u001b[0;34m(self, amt, decode_content, cache_content)\u001b[0m\n\u001b[1;32m    564\u001b[0m fp_closed \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mclosed\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m    566\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_error_catcher():\n\u001b[0;32m--> 567\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m fp_closed \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    568\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    569\u001b[0m         flush_decoder \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/code/mavemap/.venv/lib/python3.12/site-packages/urllib3/response.py:533\u001b[0m, in \u001b[0;36mHTTPResponse._fp_read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    530\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m buffer\u001b[38;5;241m.\u001b[39mgetvalue()\n\u001b[1;32m    531\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    532\u001b[0m     \u001b[38;5;66;03m# StringIO doesn't like amt=None\u001b[39;00m\n\u001b[0;32m--> 533\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mif\u001b[39;00m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp\u001b[38;5;241m.\u001b[39mread()\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.2/lib/python3.12/http/client.py:479\u001b[0m, in \u001b[0;36mHTTPResponse.read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    476\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m amt \u001b[38;5;241m>\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength:\n\u001b[1;32m    477\u001b[0m     \u001b[38;5;66;03m# clip the read to the \"end of response\"\u001b[39;00m\n\u001b[1;32m    478\u001b[0m     amt \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength\n\u001b[0;32m--> 479\u001b[0m s \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    480\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m s \u001b[38;5;129;01mand\u001b[39;00m amt:\n\u001b[1;32m    481\u001b[0m     \u001b[38;5;66;03m# Ideally, we would raise IncompleteRead if the content-length\u001b[39;00m\n\u001b[1;32m    482\u001b[0m     \u001b[38;5;66;03m# wasn't satisfied, but it might break compatibility.\u001b[39;00m\n\u001b[1;32m    483\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_close_conn()\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.2/lib/python3.12/socket.py:707\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    706\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 707\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    708\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[1;32m    709\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.2/lib/python3.12/ssl.py:1252\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1248\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1249\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1250\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   1251\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[0;32m-> 1252\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnbytes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1253\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1254\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.2/lib/python3.12/ssl.py:1104\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1102\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1103\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1104\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1105\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1106\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from dcd_mapping.align import AlignmentError, align\n",
    "from dcd_mapping.mavedb_data import get_scoreset_metadata\n",
    "\n",
    "align_results = {}\n",
    "\n",
    "for ss in tqdm(score_sets):\n",
    "    metadata = get_scoreset_metadata(ss)\n",
    "    try:\n",
    "        align_results[ss] = align(metadata, silent=True, use_cached=False)\n",
    "    except AlignmentError:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save BLAT Output\n",
    "Run the cell below to save BLAT output locally to the `analysis_files` directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "mave_blat_to_save = {}\n",
    "for ss in align_results:\n",
    "    mave_blat_to_save[ss] = align_results[ss].dict()\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(mave_blat_to_save, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Transcript and Offset Selection for MaveDB Score Sets\n",
    "In this section of the notebook, a human transcript is chosen for each protein-coding score set, and an offset is computed when the target sequence does not occur at the start of the human reference sequence. For regulatory/other non-coding score sets, a transcript is not chosen and the chromosomal sequence is selected as the reference sequence. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load BLAT output\n",
    "Run the cell below to load the BLAT output for the examined MaveDB score sets and covert alignment data from a dictionary to AlignmentResult objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for ss in mave_blat_temp:\n",
    "    align_results[ss] = AlignmentResult(**mave_blat_temp[ss])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate Transcript Mappings File\n",
    "\n",
    "Run the cell below to generate a file containing a transcript mapping for each score set. Each transcript mapping object contains the following:\n",
    "1. NM: A RefSeq transcript accession\n",
    "2. NP: A RefSeq protein sequence accession\n",
    "3. Start: An integer containing the offset for the target sequence with the respect to the selected human reference sequence\n",
    "4. Transcript Mode; The MANE status for the selected transcript (MANE Select, MANE Plus Clinical, Longest Compatible)\n",
    "5. Sequence: The translated protein reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import asyncio\n",
    "\n",
    "import nest_asyncio\n",
    "\n",
    "from dcd_mapping.resources import get_scoreset_metadata, get_scoreset_records\n",
    "from dcd_mapping.transcripts import TxSelectError, select_transcript\n",
    "\n",
    "nest_asyncio.apply()\n",
    "tx_selection = {}\n",
    "for ss in score_sets:\n",
    "    try:\n",
    "        if ss in align_results:\n",
    "            tx_selection[ss] = asyncio.run(select_transcript(metadata=\n",
    "                                                    get_scoreset_metadata(ss),\n",
    "                                                    records=get_scoreset_records(ss),\n",
    "                                                    align_result=align_results[ss],\n",
    "                                                    silent=True,))\n",
    "    except TxSelectError:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save Transcript Mappings Output\n",
    "\n",
    "Run the cell below to save the transcript_mappings file locally to the `analysis_files` directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "transcript_mappings_to_save = {}\n",
    "for ss in tx_selection:\n",
    "    if tx_selection[ss]:\n",
    "        transcript_mappings_to_save[ss] = tx_selection[ss].dict()\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(transcript_mappings_to_save, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 3: Mapping MAVE Variants using the GA4GH Variation Representation Specification (VRS)\n",
    "\n",
    "In this section of the notebook, MAVE variants are supplied to VRS, generating a pre-mapped and post-mapped computable representation for each variant. The functional effect score for each variant pair and the associated MaveDB ID are also stored in separate dictionaries."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Alignment and Transcript Selection Data\n",
    "\n",
    "Run the cell below to load alignment and transcript selection data for each score set from the `analysis_files` directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult, TxSelectResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for ss in mave_blat_temp:\n",
    "    align_results[ss] = AlignmentResult(**mave_blat_temp[ss])\n",
    "\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert MaveDB Variants to VRS Alleles\n",
    "\n",
    "Run the cell below to convert MaveDB variants to VRS objects. Each MaveDB variant has a separate pre-mapped and post-mapped list of VRS alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dcd_mapping.lookup import get_seqrepo\n",
    "from dcd_mapping.new_vrs_map import vrs_map\n",
    "from dcd_mapping.mavedb_data import get_scoreset_metadata, get_scoreset_records\n",
    "\n",
    "mave_vrs_mappings = {}\n",
    "\n",
    "for ss in align_results:\n",
    "    print(ss)\n",
    "    if ss != \"urn:mavedb:00000072-a-1\": # No RefSeq accession available for this score set\n",
    "        if ss in tx_selection:\n",
    "                mave_vrs_mappings[ss] = vrs_map(metadata=get_scoreset_metadata(ss),\n",
    "                                                align_result=align_results[ss],\n",
    "                                                records=get_scoreset_records(ss),\n",
    "                                                transcript=tx_selection[ss],\n",
    "                                                sr = get_seqrepo())\n",
    "        else:\n",
    "             mave_vrs_mappings[ss] = vrs_map(metadata=get_scoreset_metadata(ss),\n",
    "                                                align_result=align_results[ss],\n",
    "                                                records=get_scoreset_records(ss),\n",
    "                                                sr = get_seqrepo())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save VRS Mappings Dictionary\n",
    "\n",
    "Run the cell below to save the VRS mappings dictionary to `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(mave_vrs_mappings, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in VRS Mappings Dictionary and Transcript Selection Dictionary\n",
    "\n",
    "Run the cell below to read in the VRS mappings dictionary and transcript selection dictionaries from `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.transcripts import TxSelectResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"rb\") as fn:\n",
    "    mave_vrs_mappings = pickle.load(fn)\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate Annotations\n",
    "\n",
    "Run the cell below to generate annotations. These annotations are:\n",
    "1. `vrs_ref_allele_seq`: The sequence between the start and end positions indicated in the variant\n",
    "2. `hgvs`: An HGVS string describing the variant. This is only included for post-mapped variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List\n",
    "\n",
    "\n",
    "def format_start_end(ss: str, start: int, end: int) -> List[int]:\n",
    "    \"\"\"Format start and end coordinates for vrs_ref_allele_seq for known edge\n",
    "    cases\n",
    "    :param ss: score set\n",
    "    :param start: start coordinate\n",
    "    :param end: end coordinate\n",
    "    :return A list of start and end coordinates\n",
    "    \"\"\"\n",
    "    if ss.startswith(\"urn:mavedb:00000060-a-1\"):\n",
    "        # This score set set reports the entire human reference sequence as the\n",
    "        # target sequence, but the positions in the score set occur with an offset\n",
    "        # of 289\n",
    "        return [start + 289, end + 289]\n",
    "    if ss.startswith(\"urn:mavedb:00000060-a-2\"):\n",
    "        # This score set set reports the entire human reference sequence as the\n",
    "        # target sequence, but the positions in the score set occur with an offset\n",
    "        # of 331\n",
    "        return [start + 331, end + 331]\n",
    "    return [start, end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cool_seq_tool.schemas import AnnotationLayer\n",
    "\n",
    "from dcd_mapping.lookup import get_chromosome_identifier_from_vrs_id, get_seqrepo\n",
    "from dcd_mapping.utils import get_hgvs_string\n",
    "\n",
    "dp = get_seqrepo()\n",
    "\n",
    "for ss in mave_vrs_mappings:\n",
    "    print(ss)\n",
    "    for var in mave_vrs_mappings[ss].variations:\n",
    "    # Add vrs_ref_allele_seq annotation to pre-mapped variants\n",
    "        if not var:\n",
    "            continue\n",
    "        else:\n",
    "            variant_list = var.pre_mapped_variants\n",
    "        if \"members\" in variant_list:\n",
    "            for sub_var in variant_list[\"members\"]:\n",
    "                start_end = format_start_end(ss, start=sub_var[\"location\"]\n",
    "                                             [\"interval\"][\"start\"][\"value\"],\n",
    "                                             end=sub_var[\"location\"][\"interval\"]\n",
    "                                             [\"end\"][\"value\"])\n",
    "                if (ss.startswith((\"urn:mavedb:00000047\", \"urn:mavedb:00000048\",\n",
    "                                   \"urn:mavedb:00000053\", \"urn:mavedb:00000058-a-1\"))):\n",
    "                    seq = tx_selection[ss].sequence\n",
    "                    sub_var[\"vrs_ref_allele_seq\"] = seq[start_end[0]:start_end[1]]\n",
    "                else:\n",
    "                    seq = sub_var[\"location\"][\"sequence_id\"]\n",
    "                    sub_var[\"vrs_ref_allele_seq\"] = dp.get_sequence(seq, start_end[0],\n",
    "                                                                    start_end[1])\n",
    "        else:\n",
    "            start_end = format_start_end(ss, start=variant_list[\"location\"][\"interval\"]\n",
    "                                             [\"start\"][\"value\"],\n",
    "                                             end=variant_list[\"location\"][\"interval\"]\n",
    "                                             [\"end\"][\"value\"])\n",
    "            if (ss.startswith((\"urn:mavedb:00000047\", \"urn:mavedb:00000048\",\n",
    "                               \"urn:mavedb:00000053\", \"urn:mavedb:00000058-a-1\"))):\n",
    "                seq = tx_selection[ss].sequence\n",
    "                variant_list[\"vrs_ref_allele_seq\"] = seq[start_end[0]:start_end[1]]\n",
    "            else:\n",
    "                seq = variant_list[\"location\"][\"sequence_id\"]\n",
    "                variant_list[\"vrs_ref_allele_seq\"] = dp.get_sequence(seq, start_end[0],\n",
    "                                                                     start_end[1])\n",
    "\n",
    "        # Determine reference sequence\n",
    "        if var.layer == AnnotationLayer.GENOMIC:\n",
    "            if \"members\" in variant_list:\n",
    "                acc = get_chromosome_identifier_from_vrs_id(var.post_mapped_variants\n",
    "                                                            [\"members\"][0][\"location\"]\n",
    "                                                            [\"sequence_id\"])\n",
    "                acc = acc.strip(\"refseq:\")\n",
    "            else:\n",
    "                acc = get_chromosome_identifier_from_vrs_id(var.post_mapped_variants\n",
    "                                                            [\"location\"]\n",
    "                                                            [\"sequence_id\"])\n",
    "                acc = acc.strip(\"refseq:\")\n",
    "        else:\n",
    "            acc = tx_selection[ss].np\n",
    "\n",
    "        # Add vrs_ref_allele_seq annotation and hgvs string to post-mapped variants\n",
    "        variant_list = var.post_mapped_variants\n",
    "        if \"members\" in variant_list:\n",
    "            for sub_var in variant_list[\"members\"]:\n",
    "                sub_var[\"vrs_ref_allele_seq\"] = dp.get_sequence(sub_var[\"location\"][\"sequence_id\"],\n",
    "                                                                sub_var[\"location\"][\"interval\"]\n",
    "                                                                    [\"start\"][\"value\"],\n",
    "                                                                sub_var[\"location\"][\"interval\"]\n",
    "                                                                    [\"end\"][\"value\"])\n",
    "                sub_var[\"hgvs\"] = get_hgvs_string(sub_var, dp, acc)\n",
    "        else:\n",
    "            variant_list[\"vrs_ref_allele_seq\"] = dp.get_sequence(variant_list[\"location\"][\"sequence_id\"],\n",
    "                                                            variant_list[\"location\"][\"interval\"]\n",
    "                                                                [\"start\"][\"value\"],\n",
    "                                                            variant_list[\"location\"][\"interval\"]\n",
    "                                                                [\"end\"][\"value\"])\n",
    "            variant_list[\"hgvs\"] = get_hgvs_string(variant_list, dp, acc)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save Annotated VRS Mappings Dictionary\n",
    "\n",
    "Run the cell below to save the annotated VRS mappings dictionary to `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(mave_vrs_mappings, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save VRS mappings output in JSON files\n",
    "\n",
    "Run the cells below to save the VRS mappings output in a JSON file in `analysis_files/mappings`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = Path(\"analysis_files/mappings\")\n",
    "path.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult, TxSelectResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for ss in mave_blat_temp:\n",
    "    align_results[ss] = AlignmentResult(**mave_blat_temp[ss])\n",
    "\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"rb\") as fn:\n",
    "    mave_vrs_mappings = pickle.load(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dcd_mapping.utils import save_mapped_output_json\n",
    "\n",
    "for ss in mave_vrs_mappings:\n",
    "        save_mapped_output_json(ss=ss, mave_vrs_mappings=mave_vrs_mappings[ss],\n",
    "                                align_result=align_results[ss],\n",
    "                                tx_output=tx_selection.get(ss, None))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mave",
   "language": "python",
   "name": "mave"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
