{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MaveDB Mapping Analysis\n",
    "This notebook demonstrates how data from score sets in MaveDB can be mapped to human reference sequences and represented using the Variation Representation Specification (VRS) of the Global Alliance for Genomics and Health (GA4GH)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Score Set Metadata\n",
    "Run the cell below to access a dataframe containing metadata for the examined MaveDB score sets. The dataframe has 209 rows, corresponding to the number of \"human\" score sets, and 7 columns: \n",
    "1. urn: The score set identifier\n",
    "2. target_sequence: The target sequence for the score set\n",
    "3. target_sequence type: Is the target sequence a DNA or protein sequence\n",
    "4. target: The listed target for the score set (e.g. Src catalytic domain, CXCR4)\n",
    "5. assembly_id: The genome build associated with the score set\n",
    "6. uniprot_id: The Uniprot ID associated with the score set, if available\n",
    "7. target_type: The target type associated with the score set (e.g. Regulatory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from os import environ\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "environ[\"GENE_NORM_ENV\"] = \"Dev\"\n",
    "environ[\"SKIP_AWS_CONFIRMATION\"] = \"true\"\n",
    "environ[\"GENE_DYNAMO_TABLE\"] = \"gene_concepts_nonprod\"\n",
    "environ[\"UTA_DB_URL\"] = \"postgresql://uta_admin:uta@localhost:5432/uta/uta_20210129\"\n",
    "environ[\"PATH\"] = f\"/Users/rsjxa001/mave_work/dcd_mapping/notebooks/analysis:{environ['PATH']}\"\n",
    "\n",
    "mave_metadata_dat = pd.read_csv(\"analysis_files/mave_dat.csv\", index_col=[0])\n",
    "score_sets = mave_metadata_dat[\"urn\"].to_list()[:-5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Output Directory\n",
    "\n",
    "Output from this notebook will be stored in a directory named `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = Path(\"analyis_files\")\n",
    "path.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: MaveDB Metadata to BLAT Alignment Data\n",
    "In this section of the notebook, the target sequence for each score set is run through BLAT, allowing for genomic coordinates to be linked with the target sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate BLAT Output for each Score Set\n",
    "\n",
    "Run the cell below to generate BLAT alignment output for each examined score set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 9\u001b[0m\n\u001b[1;32m      7\u001b[0m metadata \u001b[39m=\u001b[39m get_scoreset_metadata(ss)\n\u001b[1;32m      8\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m----> 9\u001b[0m     align_results[ss] \u001b[39m=\u001b[39m align(metadata, silent\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m, use_cached\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m)\n\u001b[1;32m     10\u001b[0m \u001b[39mexcept\u001b[39;00m AlignmentError:\n\u001b[1;32m     11\u001b[0m     \u001b[39mpass\u001b[39;00m\n",
      "File \u001b[0;32m~/mave_work/dcd_mapping/src/dcd_mapping/align.py:320\u001b[0m, in \u001b[0;36malign\u001b[0;34m(scoreset_metadata, silent, use_cached)\u001b[0m\n\u001b[1;32m    318\u001b[0m     blat_output \u001b[39m=\u001b[39m _get_cached_blat_output(scoreset_metadata, silent)\n\u001b[1;32m    319\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 320\u001b[0m     blat_output \u001b[39m=\u001b[39m _get_blat_output(scoreset_metadata, silent)\n\u001b[1;32m    321\u001b[0m match \u001b[39m=\u001b[39m _get_best_match(blat_output, scoreset_metadata)\n\u001b[1;32m    323\u001b[0m msg \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mAlignment complete.\u001b[39m\u001b[39m\"\u001b[39m\n",
      "File \u001b[0;32m~/mave_work/dcd_mapping/src/dcd_mapping/align.py:167\u001b[0m, in \u001b[0;36m_get_blat_output\u001b[0;34m(metadata, silent)\u001b[0m\n\u001b[1;32m    165\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    166\u001b[0m     target_args \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m--> 167\u001b[0m process_result \u001b[39m=\u001b[39m _run_blat(target_args, query_file, \u001b[39m\"\u001b[39;49m\u001b[39m/dev/stdout\u001b[39;49m\u001b[39m\"\u001b[39;49m, silent)\n\u001b[1;32m    168\u001b[0m out_file \u001b[39m=\u001b[39m _write_blat_output_tempfile(process_result)\n\u001b[1;32m    170\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/mave_work/dcd_mapping/src/dcd_mapping/align.py:92\u001b[0m, in \u001b[0;36m_run_blat\u001b[0;34m(target_args, query_file, out_file, silent)\u001b[0m\n\u001b[1;32m     90\u001b[0m command \u001b[39m=\u001b[39m \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mbin_name\u001b[39m}\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m{\u001b[39;00mreference_genome_file\u001b[39m}\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m{\u001b[39;00mtarget_args\u001b[39m}\u001b[39;00m\u001b[39m -minScore=20 \u001b[39m\u001b[39m{\u001b[39;00mquery_file\u001b[39m}\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m{\u001b[39;00mout_file\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m     91\u001b[0m _logger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mRunning BLAT command: \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m, command)\n\u001b[0;32m---> 92\u001b[0m result \u001b[39m=\u001b[39m subprocess\u001b[39m.\u001b[39;49mrun(\n\u001b[1;32m     93\u001b[0m     command, shell\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m, stdout\u001b[39m=\u001b[39;49msubprocess\u001b[39m.\u001b[39;49mPIPE, stderr\u001b[39m=\u001b[39;49msubprocess\u001b[39m.\u001b[39;49mPIPE\n\u001b[1;32m     94\u001b[0m )\n\u001b[1;32m     95\u001b[0m _logger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mBLAT command finished with result \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m, result\u001b[39m.\u001b[39mreturncode)\n\u001b[1;32m     96\u001b[0m \u001b[39mif\u001b[39;00m result\u001b[39m.\u001b[39mreturncode \u001b[39m!=\u001b[39m \u001b[39m0\u001b[39m:\n",
      "File \u001b[0;32m/opt/homebrew/Cellar/python@3.11/3.11.7_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py:550\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    548\u001b[0m \u001b[39mwith\u001b[39;00m Popen(\u001b[39m*\u001b[39mpopenargs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs) \u001b[39mas\u001b[39;00m process:\n\u001b[1;32m    549\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 550\u001b[0m         stdout, stderr \u001b[39m=\u001b[39m process\u001b[39m.\u001b[39;49mcommunicate(\u001b[39minput\u001b[39;49m, timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[1;32m    551\u001b[0m     \u001b[39mexcept\u001b[39;00m TimeoutExpired \u001b[39mas\u001b[39;00m exc:\n\u001b[1;32m    552\u001b[0m         process\u001b[39m.\u001b[39mkill()\n",
      "File \u001b[0;32m/opt/homebrew/Cellar/python@3.11/3.11.7_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py:1209\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1206\u001b[0m     endtime \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m   1208\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1209\u001b[0m     stdout, stderr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_communicate(\u001b[39minput\u001b[39;49m, endtime, timeout)\n\u001b[1;32m   1210\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1211\u001b[0m     \u001b[39m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1212\u001b[0m     \u001b[39m# See the detailed comment in .wait().\u001b[39;00m\n\u001b[1;32m   1213\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/homebrew/Cellar/python@3.11/3.11.7_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py:2108\u001b[0m, in \u001b[0;36mPopen._communicate\u001b[0;34m(self, input, endtime, orig_timeout)\u001b[0m\n\u001b[1;32m   2101\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_timeout(endtime, orig_timeout,\n\u001b[1;32m   2102\u001b[0m                         stdout, stderr,\n\u001b[1;32m   2103\u001b[0m                         skip_check_and_raise\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[1;32m   2104\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(  \u001b[39m# Impossible :)\u001b[39;00m\n\u001b[1;32m   2105\u001b[0m         \u001b[39m'\u001b[39m\u001b[39m_check_timeout(..., skip_check_and_raise=True) \u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m   2106\u001b[0m         \u001b[39m'\u001b[39m\u001b[39mfailed to raise TimeoutExpired.\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m-> 2108\u001b[0m ready \u001b[39m=\u001b[39m selector\u001b[39m.\u001b[39;49mselect(timeout)\n\u001b[1;32m   2109\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_timeout(endtime, orig_timeout, stdout, stderr)\n\u001b[1;32m   2111\u001b[0m \u001b[39m# XXX Rewrite these to use non-blocking I/O on the file\u001b[39;00m\n\u001b[1;32m   2112\u001b[0m \u001b[39m# objects; they are no longer using C stdio!\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Cellar/python@3.11/3.11.7_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/selectors.py:415\u001b[0m, in \u001b[0;36m_PollLikeSelector.select\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    413\u001b[0m ready \u001b[39m=\u001b[39m []\n\u001b[1;32m    414\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 415\u001b[0m     fd_event_list \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_selector\u001b[39m.\u001b[39;49mpoll(timeout)\n\u001b[1;32m    416\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mInterruptedError\u001b[39;00m:\n\u001b[1;32m    417\u001b[0m     \u001b[39mreturn\u001b[39;00m ready\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from dcd_mapping.align import AlignmentError, align\n",
    "from dcd_mapping.resources import get_scoreset_metadata\n",
    "\n",
    "align_results = {}\n",
    "\n",
    "for ss in score_sets:\n",
    "    metadata = get_scoreset_metadata(ss)\n",
    "    try:\n",
    "        align_results[ss] = align(metadata, silent=True, use_cached=False)\n",
    "    except AlignmentError:\n",
    "        pass\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save BLAT Output\n",
    "Run the cell below to save BLAT output locally to the `analysis_files` directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'align_results' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mpickle\u001b[39;00m\n\u001b[1;32m      3\u001b[0m mave_blat_to_save \u001b[39m=\u001b[39m {}\n\u001b[0;32m----> 4\u001b[0m \u001b[39mfor\u001b[39;00m ss \u001b[39min\u001b[39;00m align_results:\n\u001b[1;32m      5\u001b[0m     mave_blat_to_save[ss] \u001b[39m=\u001b[39m align_results[ss]\u001b[39m.\u001b[39mdict()\n\u001b[1;32m      6\u001b[0m \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39manalysis_files/mave_blat_output.pickle\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mwb\u001b[39m\u001b[39m'\u001b[39m) \u001b[39mas\u001b[39;00m fn:\n",
      "\u001b[0;31mNameError\u001b[0m: name 'align_results' is not defined"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "\n",
    "mave_blat_to_save = {}\n",
    "for ss in align_results:\n",
    "    mave_blat_to_save[ss] = align_results[ss].dict()\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(mave_blat_to_save, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Transcript and Offset Selection for MaveDB Score Sets\n",
    "In this section of the notebook, a human transcript is chosen for each protein-coding score set, and an offset is computed when the target sequence does not occur at the start of the human reference sequence. For regulatory/other non-coding score sets, a transcript is not chosen and the chromosomal sequence is selected as the reference sequence. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load BLAT output\n",
    "Run the cell below to load the BLAT output for the examined MaveDB score sets and covert alignment data from a dictionary to AlignmentResult objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for ss in mave_blat_temp:\n",
    "    align_results[ss] = AlignmentResult(**mave_blat_temp[ss])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate Transcript Mappings File\n",
    "\n",
    "Run the cell below to generate a file containing a transcript mapping for each score set. Each transcript mapping object contains the following:\n",
    "1. NM: A RefSeq transcript accession\n",
    "2. NP: A RefSeq protein sequence accession\n",
    "3. Start: An integer containing the offset for the target sequence with the respect to the selected human reference sequence\n",
    "4. Transcript Mode; The MANE status for the selected transcript (MANE Select, MANE Plus Clinical, Longest Compatible)\n",
    "5. Sequence: The translated protein reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import asyncio\n",
    "\n",
    "import nest_asyncio\n",
    "\n",
    "from dcd_mapping.resources import get_scoreset_metadata, get_scoreset_records\n",
    "from dcd_mapping.transcripts import TxSelectError, select_transcript\n",
    "\n",
    "nest_asyncio.apply()\n",
    "tx_selection = {}\n",
    "for ss in score_sets:\n",
    "    try:\n",
    "        if ss in align_results:\n",
    "            tx_selection[ss] = asyncio.run(select_transcript(metadata=\n",
    "                                                    get_scoreset_metadata(ss),\n",
    "                                                    records=get_scoreset_records(ss),\n",
    "                                                    align_result=align_results[ss],\n",
    "                                                    silent=True,))\n",
    "    except TxSelectError:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save Transcript Mappings Output\n",
    "\n",
    "Run the cell below to save the transcript_mappings file locally to the `analysis_files` directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "transcript_mappings_to_save = {}\n",
    "for ss in tx_selection:\n",
    "    if tx_selection[ss]:\n",
    "        transcript_mappings_to_save[ss] = tx_selection[ss].dict()\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(transcript_mappings_to_save, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 3: Mapping MAVE Variants using the GA4GH Variation Representation Specification (VRS)\n",
    "\n",
    "In this section of the notebook, MAVE variants are supplied to VRS, generating a pre-mapped and post-mapped computable representation for each variant. The functional effect score for each variant pair and the associated MaveDB ID are also stored in separate dictionaries."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Alignment and Transcript Selection Data\n",
    "\n",
    "Run the cell below to load alignment and transcript selection data for each score set from the `analysis_files` directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult, TxSelectResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for ss in mave_blat_temp:\n",
    "    align_results[ss] = AlignmentResult(**mave_blat_temp[ss])\n",
    "\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert MaveDB Variants to VRS Alleles\n",
    "\n",
    "Run the cell below to convert MaveDB variants to VRS objects. Each MaveDB variant has a separate pre-mapped and post-mapped list of VRS alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "urn:mavedb:00000041-a-1\n",
      "urn:mavedb:00000048-a-1\n",
      "urn:mavedb:00000068-b-1\n",
      "urn:mavedb:00000045-c-1\n",
      "urn:mavedb:00000018-a-1\n",
      "urn:mavedb:00000099-a-1\n",
      "urn:mavedb:00000001-c-1\n",
      "urn:mavedb:00000049-a-3\n",
      "urn:mavedb:00000050-a-1\n",
      "urn:mavedb:00000061-i-1\n",
      "urn:mavedb:00000083-b-1\n",
      "urn:mavedb:00000094-a-5\n",
      "urn:mavedb:00000043-a-2\n",
      "urn:mavedb:00000055-0-1\n",
      "urn:mavedb:00000104-a-2\n",
      "urn:mavedb:00000005-a-6\n",
      "urn:mavedb:00000083-h-1\n",
      "urn:mavedb:00000098-a-1\n",
      "urn:mavedb:00000108-a-2\n",
      "urn:mavedb:00000001-a-4\n",
      "urn:mavedb:00000094-a-6\n",
      "urn:mavedb:00000078-a-1\n",
      "urn:mavedb:00000103-c-1\n",
      "urn:mavedb:00000043-a-1\n",
      "urn:mavedb:00000061-d-1\n",
      "urn:mavedb:00000046-a-1\n",
      "urn:mavedb:00000081-a-1\n",
      "urn:mavedb:00000097-r-1\n",
      "urn:mavedb:00000067-a-1\n",
      "urn:mavedb:00000001-c-2\n",
      "urn:mavedb:00000049-a-7\n",
      "urn:mavedb:00000060-a-2\n",
      "urn:mavedb:00000058-a-1\n",
      "urn:mavedb:00000045-k-1\n",
      "urn:mavedb:00000094-a-14\n",
      "urn:mavedb:00000069-a-2\n",
      "urn:mavedb:00000106-b-1\n",
      "urn:mavedb:00000046-a-2\n",
      "urn:mavedb:00000069-a-1\n",
      "urn:mavedb:00000034-b-1\n",
      "urn:mavedb:00000049-a-8\n",
      "urn:mavedb:00000047-c-1\n",
      "urn:mavedb:00000094-a-8\n",
      "urn:mavedb:00000041-b-1\n",
      "urn:mavedb:00000097-o-1\n",
      "urn:mavedb:00000045-a-1\n",
      "urn:mavedb:00000097-0-1\n",
      "urn:mavedb:00000097-h-1\n",
      "urn:mavedb:00000061-h-1\n",
      "urn:mavedb:00000066-a-1\n",
      "urn:mavedb:00000021-a-1\n",
      "urn:mavedb:00000108-a-3\n",
      "urn:mavedb:00000094-a-15\n",
      "urn:mavedb:00000031-a-1\n",
      "urn:mavedb:00000094-a-2\n",
      "urn:mavedb:00000094-a-4\n",
      "urn:mavedb:00000059-a-1\n",
      "urn:mavedb:00000094-a-1\n",
      "urn:mavedb:00000083-i-1\n",
      "urn:mavedb:00000045-g-1\n",
      "urn:mavedb:00000097-i-1\n",
      "urn:mavedb:00000097-w-1\n",
      "urn:mavedb:00000045-j-1\n",
      "urn:mavedb:00000103-b-1\n",
      "urn:mavedb:00000097-g-1\n",
      "urn:mavedb:00000026-a-1\n",
      "urn:mavedb:00000083-j-1\n",
      "urn:mavedb:00000049-a-4\n",
      "urn:mavedb:00000083-e-1\n",
      "urn:mavedb:00000014-a-1\n",
      "urn:mavedb:00000097-b-1\n",
      "urn:mavedb:00000097-k-1\n",
      "urn:mavedb:00000102-0-1\n",
      "urn:mavedb:00000001-b-2\n",
      "urn:mavedb:00000094-a-3\n",
      "urn:mavedb:00000061-g-1\n",
      "urn:mavedb:00000001-d-1\n",
      "urn:mavedb:00000097-p-1\n",
      "urn:mavedb:00000097-x-1\n",
      "urn:mavedb:00000055-a-1\n",
      "urn:mavedb:00000094-a-10\n",
      "urn:mavedb:00000045-i-1\n",
      "urn:mavedb:00000032-a-1\n",
      "urn:mavedb:00000101-a-1\n",
      "urn:mavedb:00000053-a-1\n",
      "urn:mavedb:00000097-v-1\n",
      "urn:mavedb:00000091-a-1\n",
      "urn:mavedb:00000048-c-1\n",
      "urn:mavedb:00000031-c-1\n",
      "urn:mavedb:00000051-b-1\n",
      "urn:mavedb:00000097-m-1\n",
      "urn:mavedb:00000001-b-1\n",
      "urn:mavedb:00000103-a-1\n",
      "urn:mavedb:00000097-f-1\n",
      "urn:mavedb:00000057-b-1\n",
      "urn:mavedb:00000083-a-1\n",
      "urn:mavedb:00000097-s-1\n",
      "urn:mavedb:00000024-a-1\n",
      "urn:mavedb:00000045-f-1\n",
      "urn:mavedb:00000072-a-1\n",
      "urn:mavedb:00000095-b-1\n",
      "urn:mavedb:00000001-a-1\n",
      "urn:mavedb:00000023-a-1\n",
      "urn:mavedb:00000049-a-2\n",
      "urn:mavedb:00000062-b-1\n",
      "urn:mavedb:00000094-a-12\n",
      "urn:mavedb:00000057-c-1\n",
      "urn:mavedb:00000094-a-7\n",
      "urn:mavedb:00000061-f-1\n",
      "urn:mavedb:00000001-a-2\n",
      "urn:mavedb:00000083-d-1\n",
      "urn:mavedb:00000003-b-2\n",
      "urn:mavedb:00000027-a-1\n",
      "urn:mavedb:00000062-a-1\n",
      "urn:mavedb:00000097-e-1\n",
      "urn:mavedb:00000005-a-5\n",
      "urn:mavedb:00000033-a-1\n",
      "urn:mavedb:00000057-d-1\n",
      "urn:mavedb:00000068-a-1\n",
      "urn:mavedb:00000049-a-6\n",
      "urn:mavedb:00000104-a-1\n",
      "urn:mavedb:00000083-c-1\n",
      "urn:mavedb:00000049-a-1\n",
      "urn:mavedb:00000028-a-1\n",
      "urn:mavedb:00000029-a-1\n",
      "urn:mavedb:00000013-b-1\n",
      "urn:mavedb:00000093-a-1\n",
      "urn:mavedb:00000001-a-3\n",
      "urn:mavedb:00000045-d-1\n",
      "urn:mavedb:00000108-a-1\n",
      "urn:mavedb:00000096-a-1\n",
      "urn:mavedb:00000003-a-2\n",
      "urn:mavedb:00000097-c-1\n",
      "urn:mavedb:00000106-a-1\n",
      "urn:mavedb:00000083-g-1\n",
      "urn:mavedb:00000036-a-2\n",
      "urn:mavedb:00000097-t-1\n",
      "urn:mavedb:00000097-l-1\n",
      "urn:mavedb:00000094-a-11\n",
      "urn:mavedb:00000027-b-1\n",
      "urn:mavedb:00000002-a-2\n",
      "urn:mavedb:00000061-c-1\n",
      "urn:mavedb:00000035-a-2\n",
      "urn:mavedb:00000031-b-1\n",
      "urn:mavedb:00000035-a-3\n",
      "urn:mavedb:00000006-a-1\n",
      "urn:mavedb:00000094-a-13\n",
      "urn:mavedb:00000045-h-1\n",
      "urn:mavedb:00000051-c-1\n",
      "urn:mavedb:00000097-u-1\n",
      "urn:mavedb:00000097-d-1\n",
      "urn:mavedb:00000036-a-1\n",
      "urn:mavedb:00000081-a-2\n",
      "urn:mavedb:00000053-a-2\n",
      "urn:mavedb:00000060-a-1\n",
      "urn:mavedb:00000013-a-1\n",
      "urn:mavedb:00000022-a-1\n",
      "urn:mavedb:00000097-z-1\n",
      "urn:mavedb:00000094-a-9\n",
      "urn:mavedb:00000045-e-1\n",
      "urn:mavedb:00000097-j-1\n",
      "urn:mavedb:00000047-b-1\n",
      "urn:mavedb:00000003-b-1\n",
      "urn:mavedb:00000054-a-1\n",
      "urn:mavedb:00000061-a-1\n",
      "urn:mavedb:00000047-a-1\n",
      "urn:mavedb:00000035-a-1\n",
      "urn:mavedb:00000015-a-1\n",
      "urn:mavedb:00000078-b-1\n",
      "urn:mavedb:00000025-a-1\n",
      "urn:mavedb:00000045-l-1\n",
      "urn:mavedb:00000065-a-1\n",
      "urn:mavedb:00000046-a-3\n",
      "urn:mavedb:00000068-c-1\n",
      "urn:mavedb:00000005-a-2\n",
      "urn:mavedb:00000061-e-1\n",
      "urn:mavedb:00000019-a-1\n",
      "urn:mavedb:00000055-b-1\n",
      "urn:mavedb:00000057-a-1\n",
      "urn:mavedb:00000097-n-1\n",
      "urn:mavedb:00000001-d-2\n",
      "urn:mavedb:00000017-a-1\n",
      "urn:mavedb:00000005-a-3\n",
      "urn:mavedb:00000083-f-1\n",
      "urn:mavedb:00000031-d-1\n",
      "urn:mavedb:00000106-c-1\n",
      "urn:mavedb:00000020-a-1\n",
      "urn:mavedb:00000007-a-1\n",
      "urn:mavedb:00000045-b-1\n",
      "urn:mavedb:00000048-b-1\n",
      "urn:mavedb:00000095-a-1\n",
      "urn:mavedb:00000016-a-1\n",
      "urn:mavedb:00000029-b-1\n",
      "urn:mavedb:00000049-a-5\n",
      "urn:mavedb:00000097-y-1\n",
      "urn:mavedb:00000005-a-1\n",
      "urn:mavedb:00000097-a-1\n",
      "urn:mavedb:00000003-a-1\n",
      "urn:mavedb:00000030-a-1\n",
      "urn:mavedb:00000005-a-4\n",
      "urn:mavedb:00000034-a-1\n",
      "urn:mavedb:00000023-a-2\n",
      "urn:mavedb:00000002-a-1\n",
      "urn:mavedb:00000107-a-1\n",
      "urn:mavedb:00000103-d-1\n",
      "urn:mavedb:00000029-a-2\n",
      "urn:mavedb:00000061-b-1\n",
      "urn:mavedb:00000097-q-1\n"
     ]
    }
   ],
   "source": [
    "from dcd_mapping.lookup import get_seqrepo\n",
    "from dcd_mapping.new_vrs_map import vrs_map\n",
    "from dcd_mapping.resources import get_scoreset_metadata, get_scoreset_records\n",
    "\n",
    "mave_vrs_mappings = {}\n",
    "\n",
    "for ss in align_results:\n",
    "    print(ss)\n",
    "    if ss != \"urn:mavedb:00000072-a-1\": # No RefSeq accession available for this score set\n",
    "        if ss in tx_selection:\n",
    "                mave_vrs_mappings[ss] = vrs_map(metadata=get_scoreset_metadata(ss),\n",
    "                                                align_result=align_results[ss],\n",
    "                                                records=get_scoreset_records(ss),\n",
    "                                                transcript=tx_selection[ss],\n",
    "                                                sr = get_seqrepo())\n",
    "        else:\n",
    "             mave_vrs_mappings[ss] = vrs_map(metadata=get_scoreset_metadata(ss),\n",
    "                                                align_result=align_results[ss],\n",
    "                                                records=get_scoreset_records(ss),\n",
    "                                                sr = get_seqrepo())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save VRS Mappings Dictionary\n",
    "\n",
    "Run the cell below to save the VRS mappings dictionary to `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(mave_vrs_mappings, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in VRS Mappings Dictionary and Transcript Selection Dictionary\n",
    "\n",
    "Run the cell below to read in the VRS mappings dictionary and transcript selection dictionaries from `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.transcripts import TxSelectResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"rb\") as fn:\n",
    "    mave_vrs_mappings = pickle.load(fn)\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate Annotations\n",
    "\n",
    "Run the cell below to generate annotations. These annotations are:\n",
    "1. `vrs_ref_allele_seq`: The sequence between the start and end positions indicated in the variant\n",
    "2. `hgvs`: An HGVS string describing the variant. This is only included for post-mapped variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List\n",
    "\n",
    "\n",
    "def format_start_end(ss: str, start: int, end: int) -> List[int]:\n",
    "    \"\"\"Format start and end coordinates for vrs_ref_allele_seq for known edge\n",
    "    cases\n",
    "    :param ss: score set\n",
    "    :param start: start coordinate\n",
    "    :param end: end coordinate\n",
    "    :return A list of start and end coordinates\n",
    "    \"\"\"\n",
    "    if ss.startswith(\"urn:mavedb:00000060-a-1\"):\n",
    "        # This score set set reports the entire human reference sequence as the\n",
    "        # target sequence, but the positions in the score set occur with an offset\n",
    "        # of 289\n",
    "        return [start + 289, end + 289]\n",
    "    if ss.startswith(\"urn:mavedb:00000060-a-2\"):\n",
    "        # This score set set reports the entire human reference sequence as the\n",
    "        # target sequence, but the positions in the score set occur with an offset\n",
    "        # of 331\n",
    "        return [start + 331, end + 331]\n",
    "    return [start, end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "urn:mavedb:00000041-a-1\n",
      "urn:mavedb:00000048-a-1\n",
      "urn:mavedb:00000068-b-1\n",
      "urn:mavedb:00000045-c-1\n",
      "urn:mavedb:00000018-a-1\n",
      "urn:mavedb:00000099-a-1\n",
      "urn:mavedb:00000001-c-1\n",
      "urn:mavedb:00000049-a-3\n",
      "urn:mavedb:00000050-a-1\n",
      "urn:mavedb:00000061-i-1\n",
      "urn:mavedb:00000083-b-1\n",
      "urn:mavedb:00000094-a-5\n",
      "urn:mavedb:00000043-a-2\n",
      "urn:mavedb:00000055-0-1\n",
      "urn:mavedb:00000104-a-2\n",
      "urn:mavedb:00000005-a-6\n",
      "urn:mavedb:00000083-h-1\n",
      "urn:mavedb:00000098-a-1\n",
      "urn:mavedb:00000108-a-2\n",
      "urn:mavedb:00000001-a-4\n",
      "urn:mavedb:00000094-a-6\n",
      "urn:mavedb:00000078-a-1\n",
      "urn:mavedb:00000103-c-1\n",
      "urn:mavedb:00000043-a-1\n",
      "urn:mavedb:00000061-d-1\n",
      "urn:mavedb:00000046-a-1\n",
      "urn:mavedb:00000081-a-1\n",
      "urn:mavedb:00000097-r-1\n",
      "urn:mavedb:00000067-a-1\n",
      "urn:mavedb:00000001-c-2\n",
      "urn:mavedb:00000049-a-7\n",
      "urn:mavedb:00000060-a-2\n",
      "urn:mavedb:00000058-a-1\n",
      "urn:mavedb:00000045-k-1\n",
      "urn:mavedb:00000094-a-14\n",
      "urn:mavedb:00000069-a-2\n",
      "urn:mavedb:00000106-b-1\n",
      "urn:mavedb:00000046-a-2\n",
      "urn:mavedb:00000069-a-1\n",
      "urn:mavedb:00000034-b-1\n",
      "urn:mavedb:00000049-a-8\n",
      "urn:mavedb:00000047-c-1\n",
      "urn:mavedb:00000094-a-8\n",
      "urn:mavedb:00000041-b-1\n",
      "urn:mavedb:00000097-o-1\n",
      "urn:mavedb:00000045-a-1\n",
      "urn:mavedb:00000097-0-1\n",
      "urn:mavedb:00000097-h-1\n",
      "urn:mavedb:00000061-h-1\n",
      "urn:mavedb:00000066-a-1\n",
      "urn:mavedb:00000021-a-1\n",
      "urn:mavedb:00000108-a-3\n",
      "urn:mavedb:00000094-a-15\n",
      "urn:mavedb:00000031-a-1\n",
      "urn:mavedb:00000094-a-2\n",
      "urn:mavedb:00000094-a-4\n",
      "urn:mavedb:00000059-a-1\n",
      "urn:mavedb:00000094-a-1\n",
      "urn:mavedb:00000083-i-1\n",
      "urn:mavedb:00000045-g-1\n",
      "urn:mavedb:00000097-i-1\n",
      "urn:mavedb:00000097-w-1\n",
      "urn:mavedb:00000045-j-1\n",
      "urn:mavedb:00000103-b-1\n",
      "urn:mavedb:00000097-g-1\n",
      "urn:mavedb:00000026-a-1\n",
      "urn:mavedb:00000083-j-1\n",
      "urn:mavedb:00000049-a-4\n",
      "urn:mavedb:00000083-e-1\n",
      "urn:mavedb:00000014-a-1\n",
      "urn:mavedb:00000097-b-1\n",
      "urn:mavedb:00000097-k-1\n",
      "urn:mavedb:00000102-0-1\n",
      "urn:mavedb:00000001-b-2\n",
      "urn:mavedb:00000094-a-3\n",
      "urn:mavedb:00000061-g-1\n",
      "urn:mavedb:00000001-d-1\n",
      "urn:mavedb:00000097-p-1\n",
      "urn:mavedb:00000097-x-1\n",
      "urn:mavedb:00000055-a-1\n",
      "urn:mavedb:00000094-a-10\n",
      "urn:mavedb:00000045-i-1\n",
      "urn:mavedb:00000032-a-1\n",
      "urn:mavedb:00000101-a-1\n",
      "urn:mavedb:00000053-a-1\n",
      "urn:mavedb:00000097-v-1\n",
      "urn:mavedb:00000091-a-1\n",
      "urn:mavedb:00000048-c-1\n",
      "urn:mavedb:00000031-c-1\n",
      "urn:mavedb:00000051-b-1\n",
      "urn:mavedb:00000097-m-1\n",
      "urn:mavedb:00000001-b-1\n",
      "urn:mavedb:00000103-a-1\n",
      "urn:mavedb:00000097-f-1\n",
      "urn:mavedb:00000057-b-1\n",
      "urn:mavedb:00000083-a-1\n",
      "urn:mavedb:00000097-s-1\n",
      "urn:mavedb:00000024-a-1\n",
      "urn:mavedb:00000045-f-1\n",
      "urn:mavedb:00000095-b-1\n",
      "urn:mavedb:00000001-a-1\n",
      "urn:mavedb:00000023-a-1\n",
      "urn:mavedb:00000049-a-2\n",
      "urn:mavedb:00000062-b-1\n",
      "urn:mavedb:00000094-a-12\n",
      "urn:mavedb:00000057-c-1\n",
      "urn:mavedb:00000094-a-7\n",
      "urn:mavedb:00000061-f-1\n",
      "urn:mavedb:00000001-a-2\n",
      "urn:mavedb:00000083-d-1\n",
      "urn:mavedb:00000003-b-2\n",
      "urn:mavedb:00000027-a-1\n",
      "urn:mavedb:00000062-a-1\n",
      "urn:mavedb:00000097-e-1\n",
      "urn:mavedb:00000005-a-5\n",
      "urn:mavedb:00000033-a-1\n",
      "urn:mavedb:00000057-d-1\n",
      "urn:mavedb:00000068-a-1\n",
      "urn:mavedb:00000049-a-6\n",
      "urn:mavedb:00000104-a-1\n",
      "urn:mavedb:00000083-c-1\n",
      "urn:mavedb:00000049-a-1\n",
      "urn:mavedb:00000028-a-1\n",
      "urn:mavedb:00000029-a-1\n",
      "urn:mavedb:00000013-b-1\n",
      "urn:mavedb:00000093-a-1\n",
      "urn:mavedb:00000001-a-3\n",
      "urn:mavedb:00000045-d-1\n",
      "urn:mavedb:00000108-a-1\n",
      "urn:mavedb:00000096-a-1\n",
      "urn:mavedb:00000003-a-2\n",
      "urn:mavedb:00000097-c-1\n",
      "urn:mavedb:00000106-a-1\n",
      "urn:mavedb:00000083-g-1\n",
      "urn:mavedb:00000036-a-2\n",
      "urn:mavedb:00000097-t-1\n",
      "urn:mavedb:00000097-l-1\n",
      "urn:mavedb:00000094-a-11\n",
      "urn:mavedb:00000027-b-1\n",
      "urn:mavedb:00000002-a-2\n",
      "urn:mavedb:00000061-c-1\n",
      "urn:mavedb:00000035-a-2\n",
      "urn:mavedb:00000031-b-1\n",
      "urn:mavedb:00000035-a-3\n",
      "urn:mavedb:00000006-a-1\n",
      "urn:mavedb:00000094-a-13\n",
      "urn:mavedb:00000045-h-1\n",
      "urn:mavedb:00000051-c-1\n",
      "urn:mavedb:00000097-u-1\n",
      "urn:mavedb:00000097-d-1\n",
      "urn:mavedb:00000036-a-1\n",
      "urn:mavedb:00000081-a-2\n",
      "urn:mavedb:00000053-a-2\n",
      "urn:mavedb:00000060-a-1\n",
      "urn:mavedb:00000013-a-1\n",
      "urn:mavedb:00000022-a-1\n",
      "urn:mavedb:00000097-z-1\n",
      "urn:mavedb:00000094-a-9\n",
      "urn:mavedb:00000045-e-1\n",
      "urn:mavedb:00000097-j-1\n",
      "urn:mavedb:00000047-b-1\n",
      "urn:mavedb:00000003-b-1\n",
      "urn:mavedb:00000054-a-1\n",
      "urn:mavedb:00000061-a-1\n",
      "urn:mavedb:00000047-a-1\n",
      "urn:mavedb:00000035-a-1\n",
      "urn:mavedb:00000015-a-1\n",
      "urn:mavedb:00000078-b-1\n",
      "urn:mavedb:00000025-a-1\n",
      "urn:mavedb:00000045-l-1\n",
      "urn:mavedb:00000065-a-1\n",
      "urn:mavedb:00000046-a-3\n",
      "urn:mavedb:00000068-c-1\n",
      "urn:mavedb:00000005-a-2\n",
      "urn:mavedb:00000061-e-1\n",
      "urn:mavedb:00000019-a-1\n",
      "urn:mavedb:00000055-b-1\n",
      "urn:mavedb:00000057-a-1\n",
      "urn:mavedb:00000097-n-1\n",
      "urn:mavedb:00000001-d-2\n",
      "urn:mavedb:00000017-a-1\n",
      "urn:mavedb:00000005-a-3\n",
      "urn:mavedb:00000083-f-1\n",
      "urn:mavedb:00000031-d-1\n",
      "urn:mavedb:00000106-c-1\n",
      "urn:mavedb:00000020-a-1\n",
      "urn:mavedb:00000007-a-1\n",
      "urn:mavedb:00000045-b-1\n",
      "urn:mavedb:00000048-b-1\n",
      "urn:mavedb:00000095-a-1\n",
      "urn:mavedb:00000016-a-1\n",
      "urn:mavedb:00000029-b-1\n",
      "urn:mavedb:00000049-a-5\n",
      "urn:mavedb:00000097-y-1\n",
      "urn:mavedb:00000005-a-1\n",
      "urn:mavedb:00000097-a-1\n",
      "urn:mavedb:00000003-a-1\n",
      "urn:mavedb:00000030-a-1\n",
      "urn:mavedb:00000005-a-4\n",
      "urn:mavedb:00000034-a-1\n",
      "urn:mavedb:00000023-a-2\n",
      "urn:mavedb:00000002-a-1\n",
      "urn:mavedb:00000107-a-1\n",
      "urn:mavedb:00000103-d-1\n",
      "urn:mavedb:00000029-a-2\n",
      "urn:mavedb:00000061-b-1\n",
      "urn:mavedb:00000097-q-1\n"
     ]
    }
   ],
   "source": [
    "from cool_seq_tool.schemas import AnnotationLayer\n",
    "\n",
    "from dcd_mapping.lookup import get_chromosome_identifier_from_vrs_id, get_seqrepo\n",
    "from dcd_mapping.utils import get_hgvs_string\n",
    "\n",
    "dp = get_seqrepo()\n",
    "\n",
    "for ss in mave_vrs_mappings:\n",
    "    print(ss)\n",
    "    for var in mave_vrs_mappings[ss].variations:\n",
    "    # Add vrs_ref_allele_seq annotation to pre-mapped variants\n",
    "        if not var:\n",
    "            continue\n",
    "        else:\n",
    "            variant_list = var.pre_mapped_variants\n",
    "        if \"members\" in variant_list:\n",
    "            for sub_var in variant_list[\"members\"]:\n",
    "                start_end = format_start_end(ss, start=sub_var[\"location\"]\n",
    "                                             [\"interval\"][\"start\"][\"value\"],\n",
    "                                             end=sub_var[\"location\"][\"interval\"]\n",
    "                                             [\"end\"][\"value\"])\n",
    "                if (ss.startswith((\"urn:mavedb:00000047\", \"urn:mavedb:00000048\",\n",
    "                                   \"urn:mavedb:00000053\", \"urn:mavedb:00000058-a-1\"))):\n",
    "                    seq = tx_selection[ss].sequence\n",
    "                    sub_var[\"vrs_ref_allele_seq\"] = seq[start_end[0]:start_end[1]]\n",
    "                else:\n",
    "                    seq = sub_var[\"location\"][\"sequence_id\"]\n",
    "                    sub_var[\"vrs_ref_allele_seq\"] = dp.get_sequence(seq, start_end[0],\n",
    "                                                                    start_end[1])\n",
    "        else:\n",
    "            start_end = format_start_end(ss, start=variant_list[\"location\"][\"interval\"]\n",
    "                                             [\"start\"][\"value\"],\n",
    "                                             end=variant_list[\"location\"][\"interval\"]\n",
    "                                             [\"end\"][\"value\"])\n",
    "            if (ss.startswith((\"urn:mavedb:00000047\", \"urn:mavedb:00000048\",\n",
    "                               \"urn:mavedb:00000053\", \"urn:mavedb:00000058-a-1\"))):\n",
    "                seq = tx_selection[ss].sequence\n",
    "                variant_list[\"vrs_ref_allele_seq\"] = seq[start_end[0]:start_end[1]]\n",
    "            else:\n",
    "                seq = variant_list[\"location\"][\"sequence_id\"]\n",
    "                variant_list[\"vrs_ref_allele_seq\"] = dp.get_sequence(seq, start_end[0],\n",
    "                                                                     start_end[1])\n",
    "\n",
    "        # Determine reference sequence\n",
    "        if var.layer == AnnotationLayer.GENOMIC:\n",
    "            if \"members\" in variant_list:\n",
    "                acc = get_chromosome_identifier_from_vrs_id(var.post_mapped_variants\n",
    "                                                            [\"members\"][0][\"location\"]\n",
    "                                                            [\"sequence_id\"])\n",
    "                acc = acc.strip(\"refseq:\")\n",
    "            else:\n",
    "                acc = get_chromosome_identifier_from_vrs_id(var.post_mapped_variants\n",
    "                                                            [\"location\"]\n",
    "                                                            [\"sequence_id\"])\n",
    "                acc = acc.strip(\"refseq:\")\n",
    "        else:\n",
    "            acc = tx_selection[ss].np\n",
    "\n",
    "        # Add vrs_ref_allele_seq annotation and hgvs string to post-mapped variants\n",
    "        variant_list = var.post_mapped_variants\n",
    "        if \"members\" in variant_list:\n",
    "            for sub_var in variant_list[\"members\"]:\n",
    "                sub_var[\"vrs_ref_allele_seq\"] = dp.get_sequence(sub_var[\"location\"][\"sequence_id\"],\n",
    "                                                                sub_var[\"location\"][\"interval\"]\n",
    "                                                                    [\"start\"][\"value\"],\n",
    "                                                                sub_var[\"location\"][\"interval\"]\n",
    "                                                                    [\"end\"][\"value\"])\n",
    "                sub_var[\"hgvs\"] = get_hgvs_string(sub_var, dp, acc)\n",
    "        else:\n",
    "            variant_list[\"vrs_ref_allele_seq\"] = dp.get_sequence(variant_list[\"location\"][\"sequence_id\"],\n",
    "                                                            variant_list[\"location\"][\"interval\"]\n",
    "                                                                [\"start\"][\"value\"],\n",
    "                                                            variant_list[\"location\"][\"interval\"]\n",
    "                                                                [\"end\"][\"value\"])\n",
    "            variant_list[\"hgvs\"] = get_hgvs_string(variant_list, dp, acc)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save Annotated VRS Mappings Dictionary\n",
    "\n",
    "Run the cell below to save the annotated VRS mappings dictionary to `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(mave_vrs_mappings, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save VRS mappings output in JSON files\n",
    "\n",
    "Run the cells below to save the VRS mappings output in a JSON file in `analysis_files/mappings`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = Path(\"analysis_files/mappings\")\n",
    "path.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult, TxSelectResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for ss in mave_blat_temp:\n",
    "    align_results[ss] = AlignmentResult(**mave_blat_temp[ss])\n",
    "\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"rb\") as fn:\n",
    "    mave_vrs_mappings = pickle.load(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dcd_mapping.utils import save_mapped_output_json\n",
    "\n",
    "for ss in mave_vrs_mappings:\n",
    "        save_mapped_output_json(ss=ss, mave_vrs_mappings=mave_vrs_mappings[ss],\n",
    "                                align_result=align_results[ss],\n",
    "                                tx_output=tx_selection.get(ss, None))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "analysis-WLxR8p6d",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
