{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MaveDB Mapping Analysis\n",
    "\n",
    "This notebook demonstrates how data from score sets in [MaveDB](https://mavedb.org/) can be mapped to human reference sequences and represented using the [Variation Representation Specification (VRS)](https://vrs.ga4gh.org/) of the [Global Alliance for Genomics and Health (GA4GH)](https://www.ga4gh.org/), as described in \"Mapping MAVE data for use in human genomics applications\" (Arbesfeld et al). After each step of the mapping pipeline, data is saved to a local [pickle](https://docs.python.org/3/library/pickle.html) checkpoint file for easy use in later steps."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "First, initialize environment parameters to enable access to required resources:* Universal Transcript Archive (UTA): see [README](https://github.com/biocommons/uta) for setup instructions. Users with access to Docker on their local devices can use the available Docker image; otherwise, start a relatively recent (version 14+) PostgreSQL instance and add data from the available database dump.\n",
    "* SeqRepo: see [README](https://github.com/biocommons/biocommons.seqrepo?tab=readme-ov-file#requirements) for setup instructions.\n",
    "* Gene Normalizer: see [documentation](https://gene-normalizer.readthedocs.io/latest/install.html) for installation instructions\n",
    "* `blat`: Must be available on the local PATH and executable by the user. Otherwise, its location can be set manually with the `BLAT_BIN_PATH` env var. See the [UCSC Genome Browser FAQ](https://genome.ucsc.edu/FAQ/FAQblat.html#blat3) for download instructions. For our experiments, we placed the binary in the same directory as these notebooks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from os import environ\n",
    "from pathlib import Path\n",
    "\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# set external resources\n",
    "environ[\"GENE_NORM_DB_URL\"] = \"postgresql://postgres@localhost:5432/gene_normalizer\"\n",
    "environ[\"UTA_DB_URL\"] = \"postgresql://uta_admin:uta@localhost:5432/uta/uta_20210129b\"\n",
    "environ[\"BLAT_BIN_PATH\"] = str(Path(\"blat\").absolute())\n",
    "environ[\"SEQREPO_ROOT_DIR\"] = \"/usr/local/share/seqrepo/2024-02-20\"  # TODO or latest/\n",
    "environ[\"DCD_MAPPING_RESOURCES_DIR\"] = str(Path(\"./mavedb_files\").absolute())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Score Set Metadata\n",
    "Run the cell below to access a dataframe containing metadata for the examined MaveDB score sets. The dataframe has 209 rows, corresponding to the number of \"human\" score sets, and 7 columns: \n",
    "1. urn: The score set identifier\n",
    "2. target_sequence: The target sequence for the score set\n",
    "3. target_sequence type: Is the target sequence a DNA or protein sequence\n",
    "4. target: The listed target for the score set (e.g. Src catalytic domain, CXCR4)\n",
    "5. assembly_id: The genome build associated with the score set\n",
    "6. uniprot_id: The Uniprot ID associated with the score set, if available\n",
    "7. target_type: The target type associated with the score set (e.g. Regulatory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mave_metadata_dat = pd.read_csv(\"analysis_files/mave_dat.csv\", index_col=[0])\n",
    "score_sets = mave_metadata_dat[\"urn\"].to_list()[:-5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Output Directory\n",
    "\n",
    "Output from this notebook will be stored in a directory named `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = Path(\"analysis_files\")\n",
    "path.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: MaveDB Metadata to BLAT Alignment Data\n",
    "In this section of the notebook, the target sequence for each score set is run through BLAT, allowing for genomic coordinates to be linked with the target sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate BLAT Output for each Score Set\n",
    "\n",
    "Run the cell below to generate BLAT alignment output for each examined score set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████| 209/209 [00:02<00:00, 85.21it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['urn:mavedb:00000105-a-1']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "from dcd_mapping.align import AlignmentError, align\n",
    "from dcd_mapping.mavedb_data import get_scoreset_metadata\n",
    "\n",
    "align_results = {}\n",
    "\n",
    "failed_scoresets = []\n",
    "for ss in tqdm(score_sets, ncols=80):\n",
    "    metadata = get_scoreset_metadata(ss)\n",
    "    try:\n",
    "        align_results[ss] = align(metadata, silent=True, use_cached=True)   # TODO delete this line, dont use cached\n",
    "        # align_results[ss] = align(metadata, silent=True, use_cached=False)    \n",
    "    except:\n",
    "        failed_scoresets.append(ss)\n",
    "print(failed_scoresets)  # 105 is expected to fail"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save BLAT Output\n",
    "Run the cell below to save BLAT output locally to the `analysis_files` directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "mave_blat_to_save = {k: v.dict() for k, v in align_results.items()}\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(mave_blat_to_save, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Transcript and Offset Selection for MaveDB Score Sets\n",
    "In this section of the notebook, a human transcript is chosen for each protein-coding score set, and an offset is computed when the target sequence does not occur at the start of the human reference sequence. For regulatory/other non-coding score sets, a transcript is not chosen and the chromosomal sequence is selected as the reference sequence. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load BLAT output\n",
    "Run the cell below to load the BLAT output for the examined MaveDB score sets and covert alignment data from a dictionary to AlignmentResult objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {k: AlignmentResult(**v) for k, v in mave_blat_temp.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate Transcript Mappings File\n",
    "\n",
    "Run the cell below to generate a file containing a transcript mapping for each score set. Each transcript mapping object contains the following:\n",
    "1. NM: A RefSeq transcript accession\n",
    "2. NP: A RefSeq protein sequence accession\n",
    "3. Start: An integer containing the offset for the target sequence with the respect to the selected human reference sequence\n",
    "4. Transcript Mode; The MANE status for the selected transcript (MANE Select, MANE Plus Clinical, Longest Compatible)\n",
    "5. Sequence: The translated protein reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████████| 209/209 [00:23<00:00,  9.08it/s]\n"
     ]
    }
   ],
   "source": [
    "import asyncio\n",
    "\n",
    "import nest_asyncio\n",
    "\n",
    "from dcd_mapping.mavedb_data import get_scoreset_metadata, get_scoreset_records\n",
    "from dcd_mapping.transcripts import TxSelectError, select_transcript\n",
    "\n",
    "nest_asyncio.apply()\n",
    "tx_selection = {}\n",
    "for ss in tqdm(score_sets):\n",
    "    if ss in align_results:\n",
    "        try:\n",
    "            tx_selection[ss] = asyncio.run(\n",
    "                select_transcript(\n",
    "                    metadata=get_scoreset_metadata(ss),\n",
    "                    records=get_scoreset_records(ss, silent=True),\n",
    "                    align_result=align_results[ss],\n",
    "                    silent=True\n",
    "                )\n",
    "            )\n",
    "        except Exception as e:\n",
    "            print(ss)\n",
    "            print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save Transcript Mappings Output\n",
    "\n",
    "Run the cell below to save the transcript_mappings file locally to the `analysis_files` directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "transcript_mappings_to_save = {}\n",
    "for ss in tx_selection:\n",
    "    if tx_selection[ss]:\n",
    "        transcript_mappings_to_save[ss] = tx_selection[ss].dict()\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(transcript_mappings_to_save, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 3: Mapping MAVE Variants using the GA4GH Variation Representation Specification (VRS)\n",
    "\n",
    "In this section of the notebook, MAVE variants are supplied to VRS, generating a pre-mapped and post-mapped computable representation for each variant. The functional effect score for each variant pair and the associated MaveDB ID are also stored in separate dictionaries."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Alignment and Transcript Selection Data\n",
    "\n",
    "Run the cell below to load alignment and transcript selection data for each score set from the `analysis_files` directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult, TxSelectResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for ss in mave_blat_temp:\n",
    "    align_results[ss] = AlignmentResult(**mave_blat_temp[ss])\n",
    "\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert MaveDB Variants to VRS Alleles\n",
    "\n",
    "Run the cell below to convert MaveDB variants to VRS objects. Each MaveDB variant has a separate pre-mapped and post-mapped list of VRS alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████| 208/208 [1:17:16<00:00, 22.29s/it]\n"
     ]
    }
   ],
   "source": [
    "from dcd_mapping.lookup import get_seqrepo\n",
    "from dcd_mapping.vrs_map import vrs_map\n",
    "from dcd_mapping.mavedb_data import get_scoreset_metadata, get_scoreset_records\n",
    "\n",
    "mave_vrs_mappings = {}\n",
    "\n",
    "for ss in tqdm(align_results):\n",
    "    if ss == \"urn:mavedb:00000072-a-1\": \n",
    "        # No RefSeq accession available for this score set        \n",
    "        continue\n",
    "    if ss in tx_selection:\n",
    "        mave_vrs_mappings[ss] = vrs_map(\n",
    "            get_seqrepo(),\n",
    "            get_scoreset_metadata(ss),\n",
    "            get_scoreset_records(ss),\n",
    "            align_results[ss],\n",
    "            tx_selection[ss],\n",
    "        )\n",
    "    else:\n",
    "        mave_vrs_mappings[ss] = vrs_map(\n",
    "            get_seqrepo(),\n",
    "            get_scoreset_metadata(ss),\n",
    "            get_scoreset_records(ss),\n",
    "            align_results[ss],\n",
    "            None  # no transcript data needed for regulatory/noncoding scoresets\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save VRS Mappings Dictionary\n",
    "\n",
    "Run the cell below to save the VRS mappings dictionary to `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(mave_vrs_mappings, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in VRS Mappings Dictionary and Transcript Selection Dictionary\n",
    "\n",
    "Run the cell below to read in the VRS mappings dictionary and transcript selection dictionaries from `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.transcripts import TxSelectResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"rb\") as fn:\n",
    "    mave_vrs_mappings = pickle.load(fn)\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate Annotations\n",
    "\n",
    "Run the cell below to generate annotations. These annotations are:\n",
    "1. `vrs_ref_allele_seq`: The sequence between the start and end positions indicated in the variant\n",
    "2. `hgvs`: An HGVS string describing the variant. This is only included for post-mapped variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List\n",
    "\n",
    "\n",
    "def format_start_end(ss: str, start: int, end: int) -> List[int]:\n",
    "    \"\"\"Format start and end coordinates for vrs_ref_allele_seq for known edge\n",
    "    cases\n",
    "    :param ss: score set\n",
    "    :param start: start coordinate\n",
    "    :param end: end coordinate\n",
    "    :return A list of start and end coordinates\n",
    "    \"\"\"\n",
    "    if ss.startswith(\"urn:mavedb:00000060-a-1\"):\n",
    "        # This score set set reports the entire human reference sequence as the\n",
    "        # target sequence, but the positions in the score set occur with an offset\n",
    "        # of 289\n",
    "        return [start + 289, end + 289]\n",
    "    if ss.startswith(\"urn:mavedb:00000060-a-2\"):\n",
    "        # This score set set reports the entire human reference sequence as the\n",
    "        # target sequence, but the positions in the score set occur with an offset\n",
    "        # of 331\n",
    "        return [start + 331, end + 331]\n",
    "    return [start, end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cool_seq_tool.schemas import AnnotationLayer\n",
    "\n",
    "from dcd_mapping.lookup import get_chromosome_identifier_from_vrs_id, get_seqrepo\n",
    "from dcd_mapping.utils import get_hgvs_string\n",
    "\n",
    "dp = get_seqrepo()\n",
    "\n",
    "for ss in mave_vrs_mappings:\n",
    "    for var in mave_vrs_mappings[ss].variations:\n",
    "    # Add vrs_ref_allele_seq annotation to pre-mapped variants\n",
    "        if not var:\n",
    "            continue\n",
    "        else:\n",
    "            variant_list = var.pre_mapped_variants\n",
    "        if \"members\" in variant_list:\n",
    "            for sub_var in variant_list[\"members\"]:\n",
    "                start_end = format_start_end(ss, start=sub_var[\"location\"]\n",
    "                                             [\"interval\"][\"start\"][\"value\"],\n",
    "                                             end=sub_var[\"location\"][\"interval\"]\n",
    "                                             [\"end\"][\"value\"])\n",
    "                if (ss.startswith((\"urn:mavedb:00000047\", \"urn:mavedb:00000048\",\n",
    "                                   \"urn:mavedb:00000053\", \"urn:mavedb:00000058-a-1\"))):\n",
    "                    seq = tx_selection[ss].sequence\n",
    "                    sub_var[\"vrs_ref_allele_seq\"] = seq[start_end[0]:start_end[1]]\n",
    "                else:\n",
    "                    seq = sub_var[\"location\"][\"sequence_id\"]\n",
    "                    sub_var[\"vrs_ref_allele_seq\"] = dp.get_sequence(seq, start_end[0],\n",
    "                                                                    start_end[1])\n",
    "        else:\n",
    "            start_end = format_start_end(ss, start=variant_list[\"location\"][\"interval\"]\n",
    "                                             [\"start\"][\"value\"],\n",
    "                                             end=variant_list[\"location\"][\"interval\"]\n",
    "                                             [\"end\"][\"value\"])\n",
    "            if (ss.startswith((\"urn:mavedb:00000047\", \"urn:mavedb:00000048\",\n",
    "                               \"urn:mavedb:00000053\", \"urn:mavedb:00000058-a-1\"))):\n",
    "                seq = tx_selection[ss].sequence\n",
    "                variant_list[\"vrs_ref_allele_seq\"] = seq[start_end[0]:start_end[1]]\n",
    "            else:\n",
    "                seq = variant_list[\"location\"][\"sequence_id\"]\n",
    "                variant_list[\"vrs_ref_allele_seq\"] = dp.get_sequence(seq, start_end[0],\n",
    "                                                                     start_end[1])\n",
    "\n",
    "        # Determine reference sequence\n",
    "        if var.layer == AnnotationLayer.GENOMIC:\n",
    "            if \"members\" in variant_list:\n",
    "                acc = get_chromosome_identifier_from_vrs_id(var.post_mapped_variants\n",
    "                                                            [\"members\"][0][\"location\"]\n",
    "                                                            [\"sequence_id\"])\n",
    "                acc = acc.strip(\"refseq:\")\n",
    "            else:\n",
    "                acc = get_chromosome_identifier_from_vrs_id(var.post_mapped_variants\n",
    "                                                            [\"location\"]\n",
    "                                                            [\"sequence_id\"])\n",
    "                acc = acc.strip(\"refseq:\")\n",
    "        else:\n",
    "            acc = tx_selection[ss].np\n",
    "\n",
    "        # Add vrs_ref_allele_seq annotation and hgvs string to post-mapped variants\n",
    "        variant_list = var.post_mapped_variants\n",
    "        if \"members\" in variant_list:\n",
    "            for sub_var in variant_list[\"members\"]:\n",
    "                sub_var[\"vrs_ref_allele_seq\"] = dp.get_sequence(sub_var[\"location\"][\"sequence_id\"],\n",
    "                                                                sub_var[\"location\"][\"interval\"]\n",
    "                                                                    [\"start\"][\"value\"],\n",
    "                                                                sub_var[\"location\"][\"interval\"]\n",
    "                                                                    [\"end\"][\"value\"])\n",
    "                sub_var[\"hgvs\"] = get_hgvs_string(sub_var, dp, acc)\n",
    "        else:\n",
    "            variant_list[\"vrs_ref_allele_seq\"] = dp.get_sequence(variant_list[\"location\"][\"sequence_id\"],\n",
    "                                                            variant_list[\"location\"][\"interval\"]\n",
    "                                                                [\"start\"][\"value\"],\n",
    "                                                            variant_list[\"location\"][\"interval\"]\n",
    "                                                                [\"end\"][\"value\"])\n",
    "            variant_list[\"hgvs\"] = get_hgvs_string(variant_list, dp, acc)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save Annotated VRS Mappings Dictionary\n",
    "\n",
    "Run the cell below to save the annotated VRS mappings dictionary to `analysis_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"wb\") as fn:\n",
    "    pickle.dump(mave_vrs_mappings, fn, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save VRS mappings output in JSON files\n",
    "\n",
    "Run the cells below to save the VRS mappings output in a JSON file in `analysis_files/mappings`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = Path(\"analysis_files/mappings\")\n",
    "path.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult, TxSelectResult\n",
    "\n",
    "with Path.open(\"analysis_files/mave_blat_output.pickle\", \"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for ss in mave_blat_temp:\n",
    "    align_results[ss] = AlignmentResult(**mave_blat_temp[ss])\n",
    "\n",
    "with Path.open(\"analysis_files/transcript_mappings.pickle\", \"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])\n",
    "\n",
    "with Path.open(\"analysis_files/mave_vrs_mappings.pickle\", \"rb\") as fn:\n",
    "    mave_vrs_mappings = pickle.load(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dcd_mapping.utils import save_mapped_output_json\n",
    "\n",
    "for ss in mave_vrs_mappings:\n",
    "        save_mapped_output_json(ss=ss, mave_vrs_mappings=mave_vrs_mappings[ss],\n",
    "                                align_result=align_results[ss],\n",
    "                                tx_output=tx_selection.get(ss, None))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mave",
   "language": "python",
   "name": "mave"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
